staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go

```diff
@@ -24,23 +24,37 @@ import (
 	"k8s.io/apiserver/pkg/storage/storagebackend"
 )
 
+var DefaultStorage = &ETCDStorage{clients: newETCD3ClientCache()}
+
 // DestroyFunc is to destroy any resources used by the storage returned in Create() together.
 type DestroyFunc func()
 
+type Storage interface {
+	Create(c storagebackend.ConfigForResource, newFunc func() runtime.Object) (storage.Interface, DestroyFunc, error)
+	CreateHealthCheck(c storagebackend.Config, stopCh <-chan struct{}) (func() error, error)
+	CreateReadyCheck(c storagebackend.Config, stopCh <-chan struct{}) (func() error, error)
+}
+
+type ETCDStorage struct {
+	clients *etcd3ClientCache
+}
```

Comment:
My rationale in refactoring this file was:

* I wanted to avoid making the cache of etcd clients a new package scoped variable.
* Currently most things in this package (e.g. `newETCD3Storage` etc) are functions rather than methods on structs, so there wasn't an obvious place (i.e. struct) into which I could inject the cache.